contracts/contracts/strategies/NativeStaking/ValidatorAccountant.sol

```diff
@@ -0,0 +1,269 @@
+    /// @dev This function could in theory be permission-less but lets allow only the Registrator (Defender Action) to call it
+    /// for now.
+    /// @return accountingValid true if accounting was successful, false if fuse is blown
+    /* solhint-enable max-line-length */
+    function doAccounting()
+        external
+        onlyRegistrator
+        whenNotPaused
+        returns (bool accountingValid)
+    {
+        // pause the accounting on failure
+        accountingValid = _doAccounting(true);
+    }
+
+    // slither-disable-start reentrancy-eth
+    function _doAccounting(bool pauseOnFail)
+        internal
+        returns (bool accountingValid)
+    {
+        if (address(this).balance < consensusRewards) {
+            return _failAccounting(pauseOnFail);
+        }
+
+        // Calculate all the new ETH that has been swept to the contract since the last accounting
+        uint256 newSweptETH = address(this).balance - consensusRewards;
+        accountingValid = true;
+
+        // send the ETH that is from fully withdrawn validators to the Vault
+        if (newSweptETH >= MAX_STAKE) {
+            uint256 fullyWithdrawnValidators;
+            // safe since MAX_STAKE is hardcoded to 32ETH
+            unchecked {
+                // explicitly cast to uint256 as we want to round to a whole number of validators
+                fullyWithdrawnValidators = uint256(newSweptETH / MAX_STAKE);
+            }
+            activeDepositedValidators -= fullyWithdrawnValidators;
+
+            uint256 wethToVault = MAX_STAKE * fullyWithdrawnValidators;
+            IWETH9(WETH_TOKEN_ADDRESS).deposit{ value: wethToVault }();
+            // slither-disable-next-line unchecked-transfer
+            IWETH9(WETH_TOKEN_ADDRESS).transfer(VAULT_ADDRESS, wethToVault);
+            _wethWithdrawnToVault(wethToVault);
+
+            emit AccountingFullyWithdrawnValidator(
+                fullyWithdrawnValidators,
+                activeDepositedValidators,
+                wethToVault
+            );
+        }
+
+        uint256 ethRemaining = address(this).balance - consensusRewards;
+        // should be less than a whole validator stake
+        require(ethRemaining < 32 ether, "unexpected accounting");
+
+        // If no Beacon chain consensus rewards swept
+        if (ethRemaining == 0) {
+            // do nothing
+            return accountingValid;
+        }
+        // Beacon chain consensus rewards swept (partial validator withdrawals)
+        else if (ethRemaining < fuseIntervalStart) {
+            // solhint-disable-next-line reentrancy
+            consensusRewards += ethRemaining;
+            emit AccountingConsensusRewards(ethRemaining);
+        }
+        // Beacon chain consensus rewards swept but also a slashed validator fully exited
+        else if (ethRemaining > fuseIntervalEnd) {
+            IWETH9(WETH_TOKEN_ADDRESS).deposit{ value: ethRemaining }();
```

Comment:
🔵 It's possible there's no clean way to do this, but it bugs me a little that we could be making two different transfers of the same coin to the same address in the same function. Feels like it's often better to do computation first, then transfers after.